core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -553,12 +562,14 @@ private[spark] class ExecutorAllocationManager(
         }
 
         // If this is the last pending task, mark the scheduler queue as empty
-        stageIdToTaskIndices.getOrElseUpdate(stageId, new mutable.HashSet[Int]) += taskIndex
+        stageIdToTaskIndices
+          .getOrElseUpdate(stageId, new mutable.HashSet[String]) += (taskIndex + "." + attemptId)
         val numTasksScheduled = stageIdToTaskIndices(stageId).size
         val numTasksTotal = stageIdToNumTasks.getOrElse(stageId, -1)
         if (numTasksScheduled == numTasksTotal) {
           // No more pending tasks for this stage
           stageIdToNumTasks -= stageId
```

Comment:
I'm not following... L557 - 585 refer to adding executors. `stageCompleted` does not even deal with executors.